### DISCUSSION TITLE
How to get the displacement of the node of the current element?

### DISCUSSION THREAD

**Discussion by Yasunaoka** [idaholab/moose#28279] [Q&A General] [answered]
### Check these boxes if you have followed the posting rules.

- [X] Q&A General is the most appropriate section for my question
- [X] I have consulted the posting Guidelines on the Discussions front page
- [X] I have searched the Discussions forum and my question has not been asked before
- [X] I have searched the MOOSE website and the documentation does not answer my question
- [X] I have formatted my post following the posting guidelines (screenshots as a last resort, triple back quotes around pasted text)

### Question

Hello Moose community!
I want to implement hourglass resistance in my app, I'm going to add a new residual to each node in computeResidualAdditional(), in this process I need to get the displacement_old and displacement_older of the four nodes of the current element, I've looked at the other code and all I found was the displacement at the qp point, and all I need is the displacement on the nodes.
It occurred to me that maybe I could get it directly from the solution vector:
```
  auto & olderglo = _sys.solutionOlder();
  auto & oldglo = _sys.solutionOld();

  std::vector<Number> old;
  std::vector<Number> older;

  oldglo.localize(old);
  olderglo.localize(older);

  // Real dx1 = old[(_current_elem->node_id(0)) * 2] - older[(_current_elem->node_id(0)) * 2];
  // Real dy1 = old[_current_elem->node_id(0) * 2 + 1] - older[_current_elem->node_id(0) * 2 + 1];
  // Real dx2 = old[(_current_elem->node_id(1)) * 2] - older[(_current_elem->node_id(1)) * 2];
  // Real dy2 = old[_current_elem->node_id(1) * 2 + 1] - older[_current_elem->node_id(1) * 2 + 1];
  // Real dx3 = old[(_current_elem->node_id(2)) * 2] - older[(_current_elem->node_id(2)) * 2];
  // Real dy3 = old[_current_elem->node_id(2) * 2 + 1] - older[_current_elem->node_id(2) * 2 + 1];
  // Real dx4 = old[(_current_elem->node_id(3)) * 2] - older[(_current_elem->node_id(3)) * 2];
  // Real dy4 = old[_current_elem->node_id(3) * 2 + 1] - older[_current_elem->node_id(3) * 2 + 1];
```
This would work with a single thread, but not with multiple threads, and I'm thinking that maybe what is returned is not the global solution vector? Do you have a better way to do this? Thanks a lot!

**Comment by GiudGiud** [comment]
Hello

Why dont you use a nodalKernel to add a residual on each node?
Does it need to be integrated over each element?

Guillaume

**Reply by Yasunaoka** [reply]
Hello!
1.I did check some of the official nodalkernel code, and it looks like there are no shape functions involved in it (from what I've seen of the code). Also, there's _qp in there, and I don't quite understand where the _qp is the integration point here (e.g. some low-order cells, where the integration point only exists inside the cell, so which cell's integration point does this _qp actually refer to?) . I'm not familiar with nodalkernel, so I didn't use nodalkernel.

2.Yes, the calculation of hourglass resistance is based on the deformed modes of the element, and according to the formula, I need: node coordinates of the current element, node displacement, node displacement_old, and the shape function (and it's gradient)
![_cgi-bin_mmwebwx-bin_webwxgetmsgimg__ MsgID=7510366257126010337 skey=@crypt_254f50e7_5513cc4b539b6a835ca999becc9d99d9 mmweb_appid=wx_webfilehelper](https://github.com/user-attachments/assets/2a619ce3-8fce-4693-a645-d77a61648ff8)
![_cgi-bin_mmwebwx-bin_webwxgetmsgimg__ MsgID=7433748225941802230 skey=@crypt_254f50e7_5513cc4b539b6a835ca999becc9d99d9 mmweb_appid=wx_webfilehelper](https://github.com/user-attachments/assets/a2277a6c-0142-4db5-9399-cdd061612a93)
![_cgi-bin_mmwebwx-bin_webwxgetmsgimg__ MsgID=1107794525454798397 skey=@crypt_254f50e7_5513cc4b539b6a835ca999becc9d99d9 mmweb_appid=wx_webfilehelper](https://github.com/user-attachments/assets/286a5c94-3c5e-4439-918d-139f1352ca64)
Under solidmechanics, can we get the node displacement in kernel? Or can this problem be solved in nodalkernel? Thank you very much for your help!

**Reply by GiudGiud** [reply]
_qp is the node index in nodal kernels. The notation is abused

**Reply by Yasunaoka** [reply]
Hi,thanks for your help!
So here is the problem:
In kernel,I can't get the node variable like displacement.
Then in nodalkernel, I can't get the element variable like the test function,although I can get the node variable.
Is there any way to solve this problem? Much appreciated!

**Reply by GiudGiud** [reply, marked as answer]
Ok so for both cases, you could use the Functor APIs in variables. 

You can see those APIs in the doxygen for variables, or you can see them in MooseFunctor.h, which has the definition of the base functor class that variables inherit

See MooseFunctorArgument.h for all the potential arguments you can use. There's nodal arguments, there s elemental arguments etc

**Comment by Yasunaoka** [comment]
Hi
```
  auto du0 = _var.getNodalValue(_current_elem->node_ref(0)) -
             _var.getNodalValueOld(_current_elem->node_ref(0));
```
problem solved! Thanks!
